lib/internal/streams/compose.js

```diff
@@ -94,63 +121,119 @@ module.exports = function compose(...streams) {
   });
 
   if (writable) {
-    d._write = function(chunk, encoding, callback) {
-      if (head.write(chunk, encoding)) {
-        callback();
-      } else {
-        ondrain = callback;
-      }
-    };
-
-    d._final = function(callback) {
-      head.end();
-      onfinish = callback;
-    };
-
-    head.on('drain', function() {
-      if (ondrain) {
-        const cb = ondrain;
-        ondrain = null;
-        cb();
-      }
-    });
+    if (isNodeStream(head)) {
+      d._write = function(chunk, encoding, callback) {
+        if (head.write(chunk, encoding)) {
+          callback();
+        } else {
+          ondrain = callback;
+        }
+      };
 
-    tail.on('finish', function() {
-      if (onfinish) {
-        const cb = onfinish;
-        onfinish = null;
-        cb();
-      }
-    });
+      d._final = function(callback) {
+        head.end();
+        onfinish = callback;
+      };
+
+      head.on('drain', function() {
+        if (ondrain) {
+          const cb = ondrain;
+          ondrain = null;
+          cb();
+        }
+      });
+    } else if (isWebStream(head)) {
+      const writable = isTransformStream(head) ? head.writable : head;
+      const writer = writable.getWriter();
+
+      d._write = async function(chunk, encoding, callback) {
+        try {
+          await writer.ready;
+          writer.write(chunk).catch(() => {});
+          callback();
+        } catch (err) {
+          callback(err);
+        }
+      };
+
+      d._final = async function(callback) {
+        try {
+          await writer.ready;
+          writer.close().catch(() => {});
+          onfinish = callback;
+        } catch (err) {
+          callback(err);
+        }
+      };
+    }
+    if (isNodeStream(tail)) {
+      tail.on('finish', function() {
```

Comment:
Nit - the two branches for node/web streams can be unified I think?

Reply:
Ah yes
Fixed!